### PR TITLE
Move plugin to src

### PR DESCRIPTION
**Pull request by zeroadam**
Move plugin to Core\Addon class

Related to #3878

### CODE REVIEW

File: src/Core/Addon.php [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+
+class Addon
+{
+	/**
+	 * @brief uninstalls an addon.
+	 *
+	 * @param string $plugin name of the addon
+	 * @return boolean
+	 */
+	public static function uninstallPlugin($plugin)
```
Comment (annando) [inline src/Core/Addon.php]: Shouldn't ```uninstall``` be enough for the function name?

File: src/Core/Addon.php [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+		}
+	}
+
+	/**
+	 * @brief installs an addon.
+	 *
+	 * @param string $plugin name of the addon
+	 * @return bool
+	 */
+	public static function installPlugin($plugin)
```
Comment (annando) [inline src/Core/Addon.php]: Same as above. ```install``` should be enough for the function name, shouldn't it?

File: src/Core/Addon.php [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+	 *   * Description: My Cool Theme
+	 * . * Version: 1.2.3
+	 *   * Author: John <profile url>
+	 *   * Maintainer: Jane <profile url>
+	 *   *
+	 * \endcode
+	 * @param string $theme the name of the theme
+	 * @return array
+	 */
+	public static function getThemeInfo($theme)
```
Comment (annando) [inline src/Core/Addon.php]: I would prefer having the theme stuff in another file.

**Comment by zeroadam** [comment]
@annando src/Core/Theme ?

File: src/Core/Addon.php (hunk 1 above)
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename any `$plugin` into `$addon`.

File: src/Core/Addon.php (hunk 2 above)
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename any `$plugin` into `$addon`.

File: src/Core/Addon.php [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+			return false;
+		}
+	}
+
+	/**
+	 * Reload all updated plugins
+	 */
+	public static function reload()
+	{
+		$plugins = Config::get('system', 'addon');
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename any `$plugins` into `$addons`.

File: src/Core/Addon.php [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+		if (strlen($plugins)) {
+			$r = q("SELECT * FROM `addon` WHERE `installed` = 1");
+			if (DBM::is_result($r)) {
+				$installed = $r;
+			} else {
+				$installed = [];
+			}
+
+			$parr = explode(',', $plugins);
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename `$parr` to `$addon_list`.

File: src/Core/Addon.php [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+
+			if (count($parr)) {
+				foreach ($parr as $pl) {
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename `$pl` to `$addon`.

File: src/Core/Addon.php [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+		}
+	}
+
+	/**
+	 * Check if addon is enabled
+	 *
+	 * @param string $plugin
+	 * @return boolean
+	 */
+	public static function isEnabled($plugin)
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename `$plugin` to `$addon`.

File: src/Core/Addon.php [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+	*   * Description: A plugin which plugs in
+	* . * Version: 1.2.3
+	*   * Author: John <profile url>
+	*   * Author: Jane <email>
+	*   *
+	*  *\endcode
+	* @param string $plugin the name of the plugin
+	* @return array with the plugin information
+	*/
+	public static function getPluginInfo($plugin)
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename function to `getInfo`.

Comment (MrPetovan) [inline src/Core/Addon.php]: Please rename `$plugin` to `$addon`.

File: src/Core/Addon.php (hunk 3 above)
Comment (MrPetovan) [reply src/Core/Addon.php]: Please move to `Friendica\Core\Theme::getInfo()`.

File: src/Core/Addon.php [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+
+	/**
+	 * Returns the theme's screenshot.
+	 *
+	 * The screenshot is expected as view/theme/$theme/screenshot.[png|jpg].
+	 *
+	 * @param sring $theme The name of the theme
+	 * @return string
+	 */
+	public static function getThemeScreenshot($theme)
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please move to `Friendica\Core\Theme::getScreenshot()`.

File: src/Core/Addon.php [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+		return(System::baseUrl() . '/images/blank.png');
+	}
+
+	// install and uninstall theme
+	/**
+	 * Uninstall theme
+	 *
+	 * @param string $theme theme
+	 */
+	public static function uninstallTheme($theme)
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please move to `Friendica\Core\Theme::uninstall()`.

File: src/Core/Addon.php [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+			$func();
+		}
+	}
+
+	/**
+	 * Install theme
+	 *
+	 * @param string $theme theme
+	 */
+	public static function installTheme($theme)
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please move to `Friendica\Core\Theme::install()`.

File: src/Core/Addon.php [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+	 *
+	 * This function search in the theme directory (and if not present in global theme directory)
+	 * if there is a directory with the file extension and  for a file with the given
+	 * filename.
+	 *
+	 * @param string $file Filename
+	 * @param string $root Full root path
+	 * @return string Path to the file or empty string if the file isn't found
+	 */
+	public static function themeInclude($file, $root = '')
```
Comment (MrPetovan) [inline src/Core/Addon.php]: Please move to `Friendica\Core\Theme::getPathForFile()`.

File: update.php [hunk 13]
```diff
@@ -79,19 +82,17 @@ function update_1189() {
...
-
-	require_once 'include/plugin.php';
-
+function update_1191()
+{
 	Config::set('system', 'maintenance', 1);
 
-	if (plugin_enabled('forumlist')) {
+	if (Addon::isEnabled('forumlist')) {
 		$plugin = 'forumlist';
```
Comment (MrPetovan) [inline update.php]: Please rename `$plugin` to `$addon`.

File: update.php [hunk 14]
```diff
@@ -79,19 +82,17 @@ function update_1189() {
...
-		$plugins = Config::get('system','addon');
+		$plugins = Config::get('system', 'addon');
```
Comment (MrPetovan) [inline update.php]: Please rename `$plugins` to `$addons`.

File: update.php [hunk 15]
```diff
@@ -79,19 +82,17 @@ function update_1189() {
...
 		$plugins_arr = [];
```
Comment (MrPetovan) [inline update.php]: Please rename `$plugins_arr` to `$addons_arr`.

File: boot.php [hunk 16]
```diff
@@ -833,7 +833,7 @@ function check_plugins(App $a)
 	if (count($installed)) {
 		foreach ($installed as $i) {
 			if (!in_array($i['name'], $plugins_arr)) {
```
Comment (MrPetovan) [inline boot.php]: Please rename `$plugins_arr` to `$addons_arr`.

File: boot.php [hunk 17]
```diff
@@ -843,12 +843,12 @@ function check_plugins(App $a)
 	if (count($plugins_arr)) {
```
Comment (MrPetovan) [inline boot.php]: Please rename `$plugins_arr` to `$addons_arr`.

File: boot.php [hunk 18]
```diff
@@ -843,12 +843,12 @@ function check_plugins(App $a)
...
 		foreach ($plugins_arr as $p) {
```
Comment (MrPetovan) [inline boot.php]: Please rename `$p` to `$addon`.

File: mod/admin.php [hunk 19]
```diff
@@ -1756,11 +1757,11 @@ function admin_page_plugins(App $a)
 			$idx = array_search($plugin, $a->plugins);
```
Comment (MrPetovan) [inline mod/admin.php]: Please rename `$plugin` to `$addon`.

Comment (MrPetovan) [inline mod/admin.php]: Please rename `$a->plugins` to `$->addons`.

File: mod/admin.php [hunk 20]
```diff
@@ -2071,7 +2072,7 @@ function admin_page_themes(App $a)
 			'$plugin' => $theme,
```
Comment (MrPetovan) [inline mod/admin.php]: Please rename `$plugin` to `$theme`.

File: mod/admin.php [hunk 21]
```diff
@@ -2103,7 +2104,7 @@ function admin_page_themes(App $a)
 
 	$plugins = [];
```
Comment (MrPetovan) [inline mod/admin.php]: Please rename `$plugins` to `$theme_list`.

File: mod/admin.php [hunk 22]
```diff
@@ -2103,7 +2104,7 @@ function admin_page_themes(App $a)
...
 	foreach ($themes as $th) {
-		$plugins[] = [$th['name'], (($th['allowed']) ? "on" : "off"), get_theme_info($th['name'])];
+		$plugins[] = [$th['name'], (($th['allowed']) ? "on" : "off"), Addon::getThemeInfo($th['name'])];
 	}
 
 	$t = get_markup_template('admin/plugins.tpl');
```
Comment (MrPetovan) [inline mod/admin.php]: Please rename `admin/plugins.tpl` to `admin/addons.tpl`.

File: mod/nodeinfo.php [hunk 23]
```diff
@@ -147,28 +147,26 @@ function nodeinfo_init(App $a) {
...
+function nodeinfo_cron()
+{
 	$a = get_app();
 
 	// If the plugin 'statistics_json' is enabled then disable it and actrivate nodeinfo.
-	if (plugin_enabled('statistics_json')) {
+	if (Addon::isEnabled('statistics_json')) {
 		Config::set('system', 'nodeinfo', true);
 
 		$plugin = 'statistics_json';
```
Comment (MrPetovan) [inline mod/nodeinfo.php]: Please rename `$plugin` to `$addon`.

File: mod/nodeinfo.php [hunk 24]
```diff
@@ -147,28 +147,26 @@ function nodeinfo_init(App $a) {
...
 		$plugins = Config::get('system', 'addon');
```
Comment (MrPetovan) [inline mod/nodeinfo.php]: Please rename `$plugins` to `$addons`.

File: mod/nodeinfo.php [hunk 25]
```diff
@@ -147,28 +147,26 @@ function nodeinfo_init(App $a) {
...
 		$plugins_arr = [];
```
Comment (MrPetovan) [inline mod/nodeinfo.php]: Please rename `$plugins_arr` to `$addons_arr`.

**Comment by zeroadam** [comment]
I rebased on develop, hopefully that completely mess everything up.

**Comment by zeroadam** [comment]
I am continuing with this work. I just needed to push so I could get it on
my node and see how things were so far. I also have about 81 files in the
addon directory that are modified as well

**Comment by zeroadam** [comment]
So, I've gotten this to the stage where everything "looks" ok (i.e. displays properly). It does however seem that enabling/disabling addons does not function. I cannot figure out why, there are no log entries or error messages, but the page just reloads. Any eyes with suggestions are appreciated!

**Comment by MrPetovan** [comment]
I'll test tonight, meanwhile you have a conflict to solve.

**Comment by MrPetovan** [comment]
Things to check:
- Is the setting correctly stored in your database?
- Is it correctly retrieved when checking for status?
- Is the naming identical in the saving/retrieving cases?

**Comment by zeroadam** [comment]
@MrPetovan it looks like the conflict is adding use Addon and maybe a re-ordering of use statements. I did a rebase, not sure why that didn't take care of that?

**Comment by MrPetovan** [comment]
Rebase effectively removes the existing commits from your branch, applies the commits from the target branch (here, develop), then applies your commits one by one.

Which means that when a conflict shows up, it’s going to come from one of your commits that are reapplied. so you’re going to see your changes as new even though you made them before the develop merge you’re rebasing on.

**Comment by zeroadam** [comment]
at this point I'm thinking to start over. I think my branch has more than one rebase in it, and things are looking quite messy (the way the branches are drawn)

**Pull request closed without merging**